File: csrc/cpp_itfs/mla/asm_mla_decode_fwd.py
from jinja2 import Template
from csrc.cpp_itfs.utils import compile_template_op, transfer_hsaco, AITER_CORE_DIR
from aiter.aot.triton_compile import compile_kernel
import triton


MD_NAME = "asm_mla_decode_fwd"
warpSize = 64
with open("asm_mla_decode_fwd.cpp.jinja", "r") as f:
    src_template = Template(f.read())


def compile(hsaco_path: str, page_size: int, q_itemsize: int, kv_itemsize: int, num_kv_splits:int, v_head_dim:int, func_name: str = None):
    bin_size, bin_data = transfer_hsaco(hsaco_path)
    triton_kernel, triton_header, triton_source = compile_kernel(f"{AITER_CORE_DIR}/aiter/ops/triton/decode_mla.py", "_fwd_kernel_stage2_asm", f"*fp32:16,*fp32:16,*bf16:16,*i32:16,i32,i32,i32,i32,i32,i32,i32,{num_kv_splits},{triton.next_power_of_2(v_head_dim)},{v_head_dim},64", "bs,nheads,1", 4, 2, "decode_mla_stage2_asm", waves_per_eu=4, kpack=2, matrix_instr_nonkdim=16)

    return compile_template_op(src_template, MD_NAME, ["../utils.h", "../../include"], [triton_source], bin_size=bin_size, bin_data=bin_data, page_size=page_size, q_itemsize=q_itemsize, kv_itemsize=kv_itemsize, triton_header=triton_header, triton_kernel=triton_kernel, func_name=func_name)


def asm_mla_decode_fwd(q,         # [num_seqs, num_heads, head_size]
                       kv_buffer,    # [num_seqs, num_heads, max_num_partitions]
                       output,
                       kv_indptr, # [num_seqs, num_heads, head_size]
                       kv_page_indices,  # [num_blocks, num_heads, head_size/x, block_size, x]
                       kv_last_page_lens,  # [num_blocks, num_heads, head_size, block_size]
                       softmax_scale=None,
                       logit_cap=0.0,
                       num_kv_splits=None,  # for experts only!!!
                       logits=None,
                       attn_lse=None,
):
    import torch
    from csrc.cpp_itfs.torch_utils import torch_to_c_types
    if not q.is_contiguous():
        raise ValueError(f"{asm_mla_decode_fwd.__name__}: only support Q.is_contiguous() for now")
    
    if q.dtype() != torch.bfloat16:
        raise ValueError(f"{asm_mla_decode_fwd.__name__}: only support dtype == torch.bfloat16 for now")
    
    num_kv_heads = kv_buffer.size(2)
    num_seqs = q.size(0)
    num_heads = q.size(1)
    head_size = q.size(2)
    page_size = kv_buffer.size(1)
    kv_split = logits.size(1)
    v_head_dim = output.size(2)

    if num_kv_heads != 1:
        raise ValueError(f"{asm_mla_decode_fwd.__name__}: only support num_kv_heads==1 for now")
    
    if head_size != kv_buffer.size(3):
        raise ValueError(f"{asm_mla_decode_fwd.__name__}: only support head_size == KV.size(3) for now")

    if logit_cap > 0:
        raise ValueError(f"{asm_mla_decode_fwd.__name__}: only support logit_cap==0 for now")
    
    if softmax_scale is None:
        softmax_scale = 1.0 / (head_size**0.5)

    if num_kv_splits is None:
        device_props = torch.cuda.get_device_properties(q.device)
        cu_num = device_props.multi_processor_count
        num_kv_splits = min(16, max(1, cu_num // num_seqs))

    func = compile(f"{AITER_CORE_DIR}/hsa/mla_stage1_a16w16_bf16.co", page_size, q.itemsize, kv_buffer.itemsize, num_kv_splits, v_head_dim)
    
    func(torch_to_c_types(q, kv_buffer, kv_indptr, kv_page_indices, kv_last_page_lens, softmax_scale, logits, attn_lse, num_seqs, num_heads, num_kv_heads, kv_split, q.stride(0), kv_buffer.stride(0), attn_lse.stride(0), attn_lse.stride(2), attn_lse.stride(1), output.stride(0), output.stride(1), torch.cuda.current_stream()))


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--hsaco_path", type=str, required=True)
    parser.add_argument("--page_size", type=int, required=True)
    parser.add_argument("--q_itemsize", type=int, required=True)
    parser.add_argument("--kv_itemsize", type=str, required=True)
    parser.add_argument("--num_kv_splits", type=int, required=True)
    parser.add_argument("--v_head_dim", type=int, required=True)
    parser.add_argument("--func_name", type=str, default=None)
    args = parser.parse_args()
    compile(**vars(args))

File: csrc/cpp_itfs/mla/asm_mla_decode_fwd.cpp
#include <fmt/core.h>
#include "asm_mla_decode_fwd.h"
#include "../utils.h"

namespace aiter{

#define MD_NAME "asm_mla_decode_fwd"


void asm_mla_decode_fwd(
                    std::optional<std::string> folder, 
                    void* q,                 //   [num_seqs, num_heads, head_size]
                    void* kv_buffer,                //   [num_page, page_size, num_kv_heads, head_size]
                    void* kv_indptr,         //   [batch_size+1]
                    void* kv_page_indices,   //   [num_page_used]
                    void* kv_last_page_lens, //   [batch_size]
                    float softmax_scale,
                    // following are output
                    void* logits, //[batch_size, num_kv_splits, num_heads, v_head_dim]
                    void* attn_lse,   //[batch_size, num_kv_splits, num_heads,  1]
                    void* output,
                    int num_seqs,
                    int num_heads,
                    int num_kv_heads,
                    int q_stride_0,
                    int kv_buffer_stride_0,
                    int attn_lse_stride_0,
                    int attn_lse_stride_1,
                    int attn_lse_stride_2,
                    int output_stride_0,
                    int output_stride_1,
                    const int page_size,
                    const std::string q_dtype,
                    const std::string kv_dtype,
                    const int num_kv_splits,
                    const int v_head_dim,
                    const hipStream_t stream) {
  std::vector<std::string> args{std::to_string(page_size), q_dtype, kv_dtype, std::to_string(num_kv_splits), std::to_string(v_head_dim)};
  std::string func_name = get_default_func_name(MD_NAME, args);
  if(!folder){
    folder = func_name;
  }
  if (not_built(folder.value())) {
    std::string cmd = fmt::format(
        R"(python3 -m csrc.cpp_itfs.mla.asm_mla_decode_fwd --hsaco_path={hsaco_path} \
                                    --page_size={page_size} \
                                    --q_dtype={q_dtype} \
                                    --kv_dtype={kv_dtype} \
                                    --num_kv_splits={num_kv_splits} \
                                    --v_head_dim={v_head_dim})",
        fmt::arg("hsaco_path", "../../../hsa/mla_stage1_a16w16_bf16.co"),
        fmt::arg("page_size", page_size),
        fmt::arg("q_dtype", q_dtype),
        fmt::arg("kv_dtype", kv_dtype),
        fmt::arg("num_kv_splits", num_kv_splits),
        fmt::arg("v_head_dim", v_head_dim));
    executeCmd(cmd);
  }
  run_lib(func_name, folder.value(), q, kv_buffer, kv_indptr, kv_page_indices, kv_last_page_lens, softmax_scale, logits, attn_lse, output, num_seqs, num_heads, num_kv_heads, q_stride_0, kv_buffer_stride_0, attn_lse_stride_0, attn_lse_stride_1, attn_lse_stride_2, output_stride_0, output_stride_1, reinterpret_cast<const void*>(stream));
}
#undef MD_NAME
}